qiskit_aer/quantum_info/states/aer_statevector.py

```diff
@@ -0,0 +1,273 @@
+
+        Returns:
+            Statevector: the output quantum state.
+
+        Raises:
+            AerError: if the operator dimension does not match the
+                         specified Statevector subsystem dimensions.
+        """
+        if qargs is None:
+            qargs = getattr(other, "qargs", None)
+
+        if isinstance(other, (QuantumCircuit, Instruction)):
+            data, aer_state = AerStatevector._from_instruction(other, self._data, **self._configs)
+            return AerStatevector(data, _aer_state=aer_state, **self._configs)
+
+        ret = copy.copy(self)
+
+        # Evolution by an Operator
+        if not isinstance(other, Operator):
+            other = Operator(other)
+
+        # check dimension
+        if self.dims(qargs) != other.input_dims():
+            raise AerError(
+                "Operator input dimensions are not equal to statevector subsystem dimensions."
+            )
+        return Statevector._evolve_operator(ret, other, qargs=qargs)
+
+    def sample_memory(self, shots, qargs=None):
+        """Sample a list of qubit measurement outcomes in the computational basis.
+
+        Args:
+            shots (int): number of samples to generate.
+            qargs (None or list): subsystems to sample measurements for,
+                                if None sample measurement of all
+                                subsystems (Default: None).
+
+        Returns:
+            np.array: list of sampled counts if the order sampled.
+
+        Additional Information:
+
+            This function *samples* measurement outcomes using the measure
+            :meth:`probabilities` for the current state and `qargs`. It does
+            not actually implement the measurement so the current state is
+            not modified.
+        """
+        if qargs is None:
+            qubits = np.arange(self._aer_state.num_qubits)
+        else:
+            qubits = np.array(qargs)
+        self._aer_state.close()
+        self._aer_state = AerState(**self._aer_state.configuration())
+        self._aer_state.initialize(self._data, copy=False)
+        samples = self._aer_state.sample_memory(qubits, shots)
+        self._data = self._aer_state.move_to_ndarray()
+        return samples
+
+    @classmethod
+    def _from_ndarray(cls, init_data, **configs):
+        aer_state = AerState()
+
+        options = AerSimulator._default_options()
+        for config_key, config_value in configs.items():
+            if options.get(config_key):
+                aer_state.configure(config_key, config_value)
+
+        if len(init_data) == 0:
+            raise AerError('initial data must be larger than 0')
+
+        num_qubits = int(np.log2(len(init_data)))
+
+        aer_state.allocate_qubits(num_qubits)
+        aer_state.initialize(data=init_data)
+
+        return aer_state.move_to_ndarray(), aer_state
+
+    @classmethod
+    def _from_instruction(cls, inst, init_data, **configs):
```

Comment:
The `init_data` here isn't super neat.  It might be a shade cleaner to get rid of that, and then in cases where you would use it, do
```python
data, state = AerStatevector._from_ndarray(init_data, config)
AerStatevector._evolve_circuit(state, inst, qubits)
```
(or whatever it is).  Then you've got a bit less duplication of the setup logic.

Reply:
`_evolve_circuit` will not release memory of `AerState` (because it does not call `move_to_ndarray()`). On the other hand, `_from_*` returns `AerState` that releases memory. Also, currently, `_from_ndarray()` copies arrays array and `_from_instruction()` does not.  I would like to keep set `init_data` though it looks wired.